### ISSUE TITLE
ScriptForCreateOrAlter still not working 

### ISSUE THREAD

**Issue by ravirambles** [microsoft/sqlmanagementobjects#126]
ScriptForCreateOrAlter flag does not seems to be generating StoreProcedure scripts with Create  Or Alter 

I am passing the following options to the scripter 

            SchemaQualify = true,
            ScriptForCreateOrAlter = true,
            IncludeDatabaseContext = true

SQL Server version being used is Microsoft SQL Server 2019 (RTM-GDR) (KB5014356) - 15.0.2095.3 (X64) 

I have found this issue which references this issue, from the looks of it this should have been fixed. 

https://github.com/microsoft/sqlmanagementobjects/issues/11

**Comment by arcdev** [comment]
I'm able to reproduce this as well.
I was reading/walking through the code looking for a workaround or a reason and it seems `StoredProcecure` only respects the `ScriptForCreateOrAlter` if `EnforceScriptingOptions` is *also* set to `true`.

see
https://github.com/microsoft/sqlmanagementobjects/blob/main/src/Microsoft/SqlServer/Management/Smo/storedprocedurebase.cs#L424

So, for a workaround, simply add the 4th line:
```
        SchemaQualify = true,
        ScriptForCreateOrAlter = true,
        IncludeDatabaseContext = true,
        EnforceScriptingOptions = true // required to respect ScriptForCreateOrAlter
```

**Comment by mattcargile** [comment]
`ScriptForAlter` has a similar issue though `EnforceScriptingOptions` doesn't fix it.

**Comment by WDpad753** [comment]
I have been working on SMO but I keep on getting create table/Views/stores procedure when it should be alter. Please find the question in detail in following link:

[SMO Question](https://stackoverflow.com/questions/77636327/how-to-get-alter-in-sql-statements-when-using-smo-class-in-c)

**Comment by shueybubbles** [comment]
@WDpad753 Per the comment above, `EnforceScriptingOptions` needs to be `true` for `ScriptForCreateOrAlter` to be applied. 

There's a [test](https://github.com/microsoft/sqlmanagementobjects/blob/0d57a1ac37776d50a6484131946ee81f9ef07a77/src/FunctionalTest/Smo/ScriptingTests/ScriptMakerTests.cs#L309) that specifies `CreateOrAlter` isn't supported for `Table` at all.

In general, SMO won't generate `Alter` scripts for unchanged objects. You can mark a text object as changed by calling `Touch()` on it. Only the text definition will be scripted, not any of its properties.

**Comment by mattcargile** [comment]
> In general, SMO won't generate `Alter` scripts for unchanged objects. You can mark a text object as changed by calling `Touch()` on it. Only the text definition will be scripted, not any of its properties.

So then how does _SQL Server Management Studio_ do its scripting? Is the method to *always* perform a `.Touch()` on an object for a "script-out"?

**Comment by shueybubbles** [comment]
yeah if you grab the IL from ObjectExplorer.dll in `Microsoft.SqlServer.Management.UI.VSIntegration.ObjectExplorer,SqlScriptMenu` class you'll find it

```C#
public static void ScriptAlter(SqlTextWriter sqlwriter, Smo.Server server, Urn[] urns, ScriptingOptions options, Func<bool> onScriptComplete)
        {
            // Note that SqlSmoObject.Touch() only works correctly for text objects
            // like views, triggers, stored procedures and functions.  For objects like
            // indexes, smo only alters properties that have actually been changed.
            // If we start scripting non-text objects for alter, we'll have to add
            // a bunch of code to SMO to make Touch() do what it needs to do for that.
            System.Diagnostics.Debug.Assert(urns != null && urns.Length == 1, "This method can only support scripting as Single-Select");
            if (urns == null || urns.Length != 1)
            {
                throw new InvalidOperationException();
            }

            // mark the smo object dirty so we can script alter
            SqlSmoObject smoObject = server.GetSmoObject(urns[0]);
            smoObject.Touch();

            // make a connection
            ScriptMaker scriptMaker = GetScriptMaker(server, options);
            scriptMaker.Preferences.ScriptForAlter = true;

            // generate the script
            StringCollection script = scriptMaker.Script(urns);
            if (onScriptComplete())
            {
                WriteStringCollection(sqlwriter, options, script);
            }
        }
```
